File: chal4.py
import string

def sc_xor(cypher_text):
    cypher_text = bytes.fromhex(cypher_text.strip())
    letters = [ord(char) for char in string.ascii_letters+" '"]

    best_score = 0
    best_plaintext = None
    best_key = None
    best_len = 0
    for key in range(0xff):
        plaintext = bytes([byte ^ key for byte in cypher_text])
        score = sum([byte in letters for byte in plaintext])/ len(plaintext)
        if score > best_score:
            best_score = score
            best_key = key
            best_plaintext = plaintext
            best_len = len(plaintext)

    return (best_score, best_key, best_plaintext, best_len)

best_score = 0
with open("4.txt", 'r') as file:
    for n, line in enumerate(file):
        score, key, pt, best_len = sc_xor(line)
        if score > best_score:
            best_score = score
            print("Line %d, Key %s, Score %f: %s" % (n, hex(key), score, pt))

File: src/chal4.c
/*
Detect single-character XOR
One of the 60-character strings in this file has been encrypted by single-character XOR.

Find it.

(Your code from #3 should help.)
*/
/* Detect single-character XOR
    
    One of the 60-character strings in this file has been encrypted by single-character XOR.

    Find it.

    (Your code from #3 should help.)

*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "crypto_tools.h"


int main() 
{
    FILE *fp;
    if ((fp = fopen("4.txt", "r")) == NULL) {
        fprintf(stderr, "ERROR: Coudln't open file!");
        exit(1);
    }

    char *cyphertext = malloc(sizeof(char) * 256);
    char *ct_bytes;

    int best_row = 0;
    struct byte_xor_chunk best = {NULL, 0, 0, 0};

    // Loop through the whole file
    for (int row=0; !feof(fp); row++) {
        // Read line from file
        if (fgets(cyphertext, 255, fp) == NULL) {
            fprintf(stderr, "ERROR: Couldn't read file!");
            exit(1);
        }

        // Strip newline
        cyphertext[strnlen(cyphertext, 256)-1] = 0;
        // Convert hex to raw bytes
        size_t ct_size = str2bytes(cyphertext, &ct_bytes);
        // Get the best key
        struct byte_xor_chunk cur = byte_xor_crack(ct_bytes, ct_size);

        // Keep track of the best score so far
        if (cur.score > best.score) {
            best = cur;
            best_row = row + 1;
        } else {
            // Throw away results that aren't the best
            // Dont forget this allocated memory!
            free(cur.plaintext);
        }
        free(ct_bytes);
    }

    printf("Line %d, Key 0x%02x, Score %f: ", best_row, best.key, best.score);
    putc('\'', stdout);
    for (int i=0; i<best.pt_size; i++) putc(best.plaintext[i], stdout);
    putc('\'', stdout);

    free(cyphertext);
    return 0;
}
